File: backend/app/schemas/progress.py
"""
Pydantic schemas for reading progress.
"""
from datetime import datetime
from pydantic import BaseModel, Field, field_validator

from app.schemas.base import BaseSchema


class ProgressUpdate(BaseModel):
    """Schema for updating reading progress."""
    progress_percent: int = Field(..., ge=0, le=100)
    last_position: str | None = Field(None, max_length=100)


class ProgressResponse(BaseSchema):
    """Schema for reading progress response."""
    book_id: int
    progress_percent: int
    last_position: str | None
    last_read_at: datetime | None


class ProgressListResponse(BaseModel):
    """Schema for list of reading progress."""
    items: list[ProgressResponse]

File: frontend/src/types/progress.ts
export interface ReadingProgress {
  book_id: number;
  progress_percent: number;
  last_position: string | null;
  last_read_at: string | null;
}

export interface ProgressUpdate {
  progress_percent: number;
  last_position?: string;
}

export interface ProgressList {
  items: ReadingProgress[];
}